src/zcap.rs

```diff
@@ -0,0 +1,473 @@
+        };
+        let mut loader = StaticLoader;
+        json_to_dataset(&json, more_contexts.as_ref(), false, None, &mut loader).await
+    }
+}
+
+// limited initial definition of a ZCAP Invocation, generic over Action
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Invocation<A> {
+    #[serde(rename = "@context")]
+    pub context: Contexts,
+    pub id: URI,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub capability_action: Option<A>,
+    // This field is populated only when using
+    // embedded proofs such as LD-PROOF
+    //   https://w3c-ccg.github.io/ld-proofs/
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proof: Option<Proof>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub property_set: Option<Map<String, Value>>,
+}
+
+impl<A> Invocation<A>
+where
+    A: Serialize + Send + Sync + Clone,
+{
+    pub async fn verify<C>(
+        &self,
+        _options: Option<LinkedDataProofOptions>,
+        resolver: &dyn DIDResolver,
+        // TODO make this a list for delegation chains
+        target_capability: &Delegation<A, C>,
+    ) -> VerificationResult
+    where
+        C: Serialize + Send + Sync + Clone,
+    {
+        match &self.proof {
+            None => VerificationResult::error("No applicable proof"),
+            Some(proof) => {
+                let mut result = proof.verify(self, resolver).await;
+                match (
+                    // get cap id from proof extra properties
+                    proof
+                        .property_set
+                        .as_ref()
+                        .and_then(|ps| ps.get("capability").map(|s| s.clone()))
+                        .and_then(|v| match v {
+                            Value::String(id) => Some(id),
+                            _ => None,
+                        }),
+                    &target_capability.id,
+                ) {
+                    (Some(ref id), URI::String(ref t_id)) => {
+                        // ensure proof target cap ID and given
+                        if id != t_id {
+                            result
+                                .errors
+                                .push("Target Capability IDs dont match".into())
+                        };
+                    }
+                    _ => result
+                        .errors
+                        .push("Missing proof target capability ID".into()),
+                };
+                // if there are invokers listed in the target, ensure the inoker here is the right one
+                if let (Some(URI::String(ref invoker)), Some(ref delegatee)) =
+                    (&target_capability.invoker, &proof.verification_method)
+                {
+                    if invoker != delegatee {
+                        result.errors.push("Incorrect Invoker".into());
+                    }
+                };
```

Comment:
I'm wondering about the cases not checked:
- No invoker: doesn't this mean there is no cryptographic material to verify against (since capability chain is not yet supported), so an error should result?
- No verification method: probably `proof.verify` already failed in this case. Some older proof types like RsaSignature2016, appear to use `creator` instead of `verificationMethod`, but we aren't supporting those currently and I'm not sure how we would. Would it be safer to error in this case anyway?

Reply:
for case 1, invoker is optional, an empty value implies that anyone should be able to invoke the cap and that the value should not be checked. for case 2, yeah it would probably be safer to error here as the invocation should always list the invoker/proof creator

Reply:
> an empty value implies that anyone should be able to invoke the cap and that the value should not be checked

How do you find that? From the spec I gather that no `invoker` means that no new entity is granted authority; and that a capability or invocation is required to have a proof created by an entity with authority (either an `invoker` entity or `capabilityDelegation` object property of the target/root capability/document).
e.g. https://w3c-ccg.github.io/zcap-ld/#capabilities "Every capability document, except for the target, MUST have: [...] a proof field, which MUST sign the document with Linked Data Proofs by cryptographic material which has already been granted authority on the chain (either being cryptographic material from the capabilityDelegation property of the target, or from an invoker previously granted authority on the chain."
https://w3c-ccg.github.io/zcap-ld/#invocation "In the context of ZCAP-LD, an invocation consists of a linked data object which MUST have a proof property [...] the proof must validate against cryptographic material granted authority by the invoker field from the capability chain."
and the note "Why don't all child capabilities require a invoker property? The reason is that it is perfectly reasonable for an entity which has authority to a capability to give an attenuated capability to *itself* to be used in a specific context. [....]"

Reply:
> an empty value implies that anyone should be able to invoke the cap and that the value should not be checked

This sounds like a bad default that could lead to the same `"alg": "none"` vulns we've seen with JWTs. Could we instead fail violently, and have a more specific indicator that there is no invoker if that is the desired behavior of the capability issuer? I'm okay deviating from the spec and advocating for this default at the CCG.

https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/

Reply:
@clehner is right, a missing `invoker` signals that the capability is scoped to the invokers authorized in the parent delegation chain. a missing verification method will cause any verification to fail, but how might we signal that any `capabilityInvocation` VM can can invoke the cap? `invoker: "*"`?

Reply:
I think that's a good explicit one. I would also be okay with the non-URI out-of-band indicators `"any"` and `"parent"`.

Reply:
Handling missing verification method is fixed in e5173435e3adbc7a99ff75b9f9e402b962cf43c7. The comment could be updated to reflect the change.

The target's capabilityDelegation verification methods should also be considered authorized - i.e. from https://w3c-ccg.github.io/zcap-ld/#invocation : "The target's capabilityDelegation cryptographic material is marked in the initial set of authority". But not doing this is to "fail closed", so does not need to block this PR.

Allowing the capability as like a bearer token could be an opportunity to propose an addition to the `zcap-ld` spec. But looking through existing issues, I seem to find arguments against that idea:
- https://github.com/w3c-ccg/zcap-ld/issues/6#issuecomment-379830751 "It is preferable from a security perspective for tokens to be tied to holders".
- https://github.com/w3c-ccg/zcap-ld/issues/6#issuecomment-379865174 " While there are such things as bearer based capabilities (Macaroons is one), ld-ocap is not such a system." [ld-ocap was the [previous name](https://github.com/w3c-ccg/zcap-ld/commit/bcf00d2fbe9a2513c39a30808b2bcf2bb6e1dce4) of zcap-ld]

Reply:
Continuing this thread at https://github.com/spruceid/ssi/pull/189/files#r661671329